### ISSUE TITLE
Crash in -[AFURLSessionManager uploadTaskWithStreamedRequest:progress:completionHandler:]

### ISSUE THREAD

**Issue by blixt** [AFNetworking/AFNetworking#2597]
This happened when I tried to upload a file with a reference to an `NSProgress` pointer to track progress (the pointer was set to `nil` at the time).

This is the code that makes the call into AFNetworking:

```obj-c
- (void)uploadAudioURL:(NSURL *)audioURL
                  path:(NSString *)path
            parameters:(NSDictionary *)parameters
              progress:(NSProgress *__autoreleasing *)progress
               success:(SuccessCallback)success
               failure:(FailureCallback)failure
{
    UIApplication *application = [UIApplication sharedApplication];
    UIBackgroundTaskIdentifier task = [application beginBackgroundTaskWithExpirationHandler:^{
        NSLog(@"WARNING: Failed to upload audio in time.");
        [application endBackgroundTask:task];
    }];

    NSString *url = [[NSURL URLWithString:path relativeToURL:self.manager.baseURL] absoluteString];

    void (^block)(id<AFMultipartFormData>) = ^void(id<AFMultipartFormData> formData) {
        // Add the audio data to the request.
        [formData appendPartWithFileURL:audioURL name:@"audio" error:nil];
    };

    NSMutableURLRequest *request = [self.manager.requestSerializer multipartFormRequestWithMethod:@"POST"
                                                                                        URLString:url
                                                                                       parameters:parameters
                                                                        constructingBodyWithBlock:block
                                                                                            error:nil];

    AFURLSessionManager *manager = [[AFURLSessionManager alloc] initWithSessionConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    NSURLSessionUploadTask *upload = [manager uploadTaskWithStreamedRequest:request progress:progress completionHandler:^(NSURLResponse *response, id data, NSError *error) {
        if (error) {
            NSLog(@"WARNING: Failed to upload audio (%@)", error.localizedDescription);
            NSInteger statusCode = -1;
            if ([response isKindOfClass:[NSHTTPURLResponse class]]) {
                statusCode = ((NSHTTPURLResponse *)response).statusCode;
            }
            failure(statusCode, error);
        } else {
            NSLog(@"Completed upload of audio");
            success(data);
        }
        [application endBackgroundTask:task];
    }];

    [upload resume];
}
```

This is the exception: `EXC_BAD_ACCESS KERN_INVALID_ADDRESS at 0x2b08060143070600`

Call stack:

```
Thread : Crashed: com.apple.main-thread
0  libobjc.A.dylib                0x00000001984b3bdc objc_msgSend + 28
1  Foundation                     0x00000001882fc2f4 NSKeyValuePushPendingNotificationPerThread + 340
2  Foundation                     0x00000001882e94fc NSKeyValueWillChange + 560
3  Foundation                     0x00000001882d3188 -[NSObject(NSKeyValueObserverNotification) willChangeValueForKey:] + 224
4  Foundation                     0x00000001884b3ec4 -[NSProgress _setValueForKeys:settingBlock:] + 292
5  Foundation                     0x00000001884b4230 -[NSProgress setTotalUnitCount:] + 244
6  Foundation                     0x00000001884b2e88 +[NSProgress progressWithTotalUnitCount:] + 116
7  Roger                          0x0000000100081178 -[AFURLSessionManagerTaskDelegate init] (AFURLSessionManager.m:128)
8  Roger                          0x0000000100083964 -[AFURLSessionManager addDelegateForUploadTask:progress:completionHandler:] (AFURLSessionManager.m:464)
9  Roger                          0x00000001000851b4 -[AFURLSessionManager uploadTaskWithStreamedRequest:progress:completionHandler:] (AFURLSessionManager.m:658)
10 Roger                          0x0000000100060e94 -[RogerAPI uploadAudioURL:path:parameters:progress:success:failure:] (RogerAPI.m:266)
11 Roger                          0x00000001000603f8 -[RogerAPI sendMessageToAccount:audioURL:progress:success:failure:] (RogerAPI.m:176)
12 Roger                          0x0000000100062768 -[RogerContact messageWithAudioURL:progress:callback:] (RogerContact.m:127)
13 Roger                          0x0000000100052704 -[RogerTableViewCell endRecording] (RogerTableViewCell.m:216)
14 Roger                          0x0000000100059cb0 -[InboxTableViewController handleLongPress:] (InboxTableViewController.m:157)
15 UIKit                          0x000000018be6f648 _UIGestureRecognizerSendActions + 276
16 UIKit                          0x000000018bd07d30 -[UIGestureRecognizer _updateGestureWithEvent:buttonEvent:] + 580
17 UIKit                          0x000000018c175574 ___UIGestureRecognizerUpdate_block_invoke662 + 60
18 UIKit                          0x000000018bccc38c _UIGestureRecognizerRemoveObjectsFromArrayAndApplyBlocks + 292
19 UIKit                          0x000000018bcca738 _UIGestureRecognizerUpdate + 2504
20 UIKit                          0x000000018bd05ed8 -[UIWindow _sendGesturesForEvent:] + 1044
21 UIKit                          0x000000018bd05578 -[UIWindow sendEvent:] + 660
22 UIKit                          0x000000018bcd8e60 -[UIApplication sendEvent:] + 264
23 UIKit                          0x000000018bf7846c _UIApplicationHandleEventFromQueueEvent + 15220
24 UIKit                          0x000000018bcd73d0 _UIApplicationHandleEventQueue + 1716
25 CoreFoundation                 0x0000000187482d34 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
26 CoreFoundation                 0x0000000187481fd8 __CFRunLoopDoSources0 + 264
27 CoreFoundation                 0x0000000187480088 __CFRunLoopRun + 712
28 CoreFoundation                 0x00000001873ad1f4 CFRunLoopRunSpecific + 396
29 GraphicsServices               0x00000001907cf6fc GSEventRunModal + 168
30 UIKit                          0x000000018bd3e10c UIApplicationMain + 1488
31 Roger                          0x0000000100061624 main (main.m:6)
32 libdyld.dylib                  0x0000000198b1ea08 start + 4
```

**Comment by mattt** [comment]
Why are you creating a manager in this method? Because nothing else owns it, there are no guarantees of it sticking around beyond the scope of this method...

**Comment by blixt** [comment]
@mattt I haven't worked directly with `NSURLSession`, but I thought that once the task has been kicked off it's managed by the OS, and the completion block will be retained until the task finishes or the app shuts down? Is there more being managed by the `AFURLSessionManager` instance?

**Comment by mattt** [comment]
@blixt Aspects of the tasks, like progress, are still managed by the manager. Even the completion block of tasks created by managers are owned by the manager, not the task or the underlying session.

**Comment by blixt** [comment]
Ah okay, so the solution for me is simply to maintain a reference to the manager for at least the period of the upload task? If so, sorry for the misunderstanding and feel free to close this issue. :)

**Comment by mattt** [comment]
@blixt It's the most likely cause. Give that a try and let me know if that solves your issue.

**Comment by blixt** [comment]
I've yet to run into the exact same crash again, but I have however run into a very similar crash in the code that downloads audio (the other crash was in the upload part). This part of the code always had a retained session manager though, so that shouldn't be the reason.

What's strange to me is that in this case I'm passing in `nil` for the progress which I believe is the correct way to ignore it, but the crash is once again related to `NSProgress`. Am I doing something stupid here? Note that `downloadManager` is not the same instance as the one I'm using for uploading (because they're part of different classes), so I have two independent `AFURLSessionManager` instances. Is that ok?

Here's the stack:

```
Thread : Crashed: com.apple.main-thread
0  libobjc.A.dylib                0x0000000192ef3bd0 objc_msgSend + 16
1  Foundation                     0x0000000182d902f4 NSKeyValuePushPendingNotificationPerThread + 340
2  Foundation                     0x0000000182d7d4fc NSKeyValueWillChange + 560
3  Foundation                     0x0000000182d67188 -[NSObject(NSKeyValueObserverNotification) willChangeValueForKey:] + 224
4  Foundation                     0x0000000182f47ec4 -[NSProgress _setValueForKeys:settingBlock:] + 292
5  Foundation                     0x0000000182f48230 -[NSProgress setTotalUnitCount:] + 244
6  Foundation                     0x0000000182f46e88 +[NSProgress progressWithTotalUnitCount:] + 116
7  Roger                          0x00000001000d50cc -[AFURLSessionManagerTaskDelegate init] (AFURLSessionManager.m:128)
8  Roger                          0x00000001000d7c98 -[AFURLSessionManager addDelegateForDownloadTask:progress:destination:completionHandler:] (AFURLSessionManager.m:503)
9  Roger                          0x00000001000d937c -[AFURLSessionManager downloadTaskWithRequest:progress:destination:completionHandler:] (AFURLSessionManager.m:675)
10 Roger                          0x00000001000bbf84 -[RogerAudio cacheRemoteAudioURL:] (RogerAudio.m:131)
11 Roger                          0x00000001000bc8bc -[RogerAudio getDataForRemoteURL:] (RogerAudio.m:234)
12 Roger                          0x00000001000bbba8 -[RogerAudio beginPlayingURL:error:callback:] (RogerAudio.m:88)
13 Roger                          0x00000001000a77ac -[RogerTableViewCell beginPlaying] (RogerTableViewCell.m:346)
14 Roger                          0x00000001000ff358 -[FSCSwipeCell setOffset:duration:completion:] (FSCSwipeCell.m:123)
15 Roger                          0x00000001001001e0 -[FSCSwipeCell cellWasSwiped:] (FSCSwipeCell.m:255)
16 UIKit                          0x0000000186903648 _UIGestureRecognizerSendActions + 276
17 UIKit                          0x000000018679bd30 -[UIGestureRecognizer _updateGestureWithEvent:buttonEvent:] + 580
18 UIKit                          0x0000000186c09574 ___UIGestureRecognizerUpdate_block_invoke662 + 60
19 UIKit                          0x000000018676038c _UIGestureRecognizerRemoveObjectsFromArrayAndApplyBlocks + 292
20 UIKit                          0x000000018675e738 _UIGestureRecognizerUpdate + 2504
21 UIKit                          0x0000000186799ed8 -[UIWindow _sendGesturesForEvent:] + 1044
22 UIKit                          0x0000000186799578 -[UIWindow sendEvent:] + 660
23 UIKit                          0x000000018676ce60 -[UIApplication sendEvent:] + 264
24 UIKit                          0x0000000186a0c46c _UIApplicationHandleEventFromQueueEvent + 15220
25 UIKit                          0x000000018676b3d0 _UIApplicationHandleEventQueue + 1716
26 CoreFoundation                 0x0000000181f16d34 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
27 CoreFoundation                 0x0000000181f15fd8 __CFRunLoopDoSources0 + 264
28 CoreFoundation                 0x0000000181f14088 __CFRunLoopRun + 712
29 CoreFoundation                 0x0000000181e411f4 CFRunLoopRunSpecific + 396
30 GraphicsServices               0x000000018b26b6fc GSEventRunModal + 168
31 UIKit                          0x00000001867d210c UIApplicationMain + 1488
32 Roger                          0x00000001000b5578 main (main.m:6)
33 libdyld.dylib                  0x000000019355ea08 start + 4
```

And here's the source that kicks off the download:

```obj-c
- (BOOL)cacheRemoteAudioURL:(NSURL *)url {
    @synchronized(self) {
        if ([self.pendingDownloads containsObject:url]) {
            return NO;
        }

        NSURL *destination = [self getLocalURLForURL:url];
        if ([[NSFileManager defaultManager] fileExistsAtPath:[destination path]]) {
            return YES;
        }

        [self.pendingDownloads addObject:url];
    }

#ifdef DEBUG
    NSLog(@"Downloading %@", url.lastPathComponent);
#endif

    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    NSURLSessionDownloadTask *task;
    task = [self.downloadManager downloadTaskWithRequest:request progress:nil destination:nil completionHandler:^(NSURLResponse *response, NSURL *filePath, NSError *error) {
       [self.pendingDownloads removeObject:url];
        dispatch_async(dispatch_get_main_queue(),^{
            [[NSNotificationCenter defaultCenter] postNotificationName:RogerAudioCachedNotification object:nil userInfo:@{@"url": url}];
        });
    }];
    [task resume];
    return NO;
}
```

**Comment by jojojames** [comment]
+1 seeing a crash right around here.

- (instancetype)init {
    self = [super init];
    if (!self) {
        return nil;
    }

    self.mutableData = [NSMutableData data];

    NSLog(@"AFURLSessionManagerTaskDelegate-before setting progress"); <-- crashes here
    self.progress = [NSProgress progressWithTotalUnitCount:0];
    NSLog(@"AFURLSessionManagerTaskDelegate-after setting progress"); <-- never reaches here
    return self;
}

Code which leads to that.

-(void)sendMediaRequestForMedia:(id<MediaDownloadHelper>)mediaHelper {
    [mediaHelper saveContext:DownloadMediaContext];
    
    NSString* address = [mediaHelper mediaAddress];
    if (!address) {
        return;
    }
    
    NSURL *url = [NSURL URLWithString:address];
    NSMutableURLRequest* request = [[NSMutableURLRequest alloc] initWithURL:url];
    
    NSProgress* progress = nil;
    
    NSURLSessionDownloadTask *downloadTask = [self.downloadSessionManager
                                              downloadTaskWithRequest:request
                                              progress:&progress
                                              destination:^NSURL *(NSURL *targetPath, NSURLResponse *response) {
                                                  
        NSURL *documentsDirectoryURL = [[NSFileManager defaultManager] URLForDirectory:NSDocumentDirectory inDomain:NSUserDomainMask appropriateForURL:nil create:NO error:nil];
        return [documentsDirectoryURL URLByAppendingPathComponent:[response suggestedFilename]];
                                                  
    } completionHandler:^(NSURLResponse *response, NSURL *filePath, NSError *error) {
        
        BOOL success = !error;
        if (success) {
            NSLog(@"File downloaded to: %@", filePath);
        } else {
            NSLog(@"File failed to download to: %@, error is %@", filePath, error);
        }
        
        [progress removeObserver:mediaHelper forKeyPath:self.progressFractionCompleted];
        [mediaHelper end:success atFilePath:filePath];
    }];
    
    [mediaHelper begin];
    [downloadTask resume];
    [progress addObserver:mediaHelper forKeyPath:self.progressFractionCompleted options:NSKeyValueObservingOptionNew context:DownloadMediaContext];
}

**Comment by cesarvilera** [comment]
I'm having the same problem when using backgroundSession... if I change it for defaultSession the problem stops... 

I'm also having another problem with this method, the content-length sent by the app is a little bit smaller than the real one.

App Content-Length: 198379 (I'm using OWASP ZAP as a proxy)
Real Content-Length: 204169

Here is my code.

```objective-c
-(void)uploadFile:(Files *)file start:(BOOL)start {
    NSString *URLString = [NSString stringWithFormat:@"%@%@",self.backgroundNetwork.baseURL,[Files getURL]];
    NSData *fileData = UIImageJPEGRepresentation(file.files_images.image, 1.0);
    NSMutableURLRequest *request;
    request = [[AFJSONRequestSerializer serializer] multipartFormRequestWithMethod:@"POST"
                                                                         URLString:URLString
                                                                        parameters:@{@"name":file.name}
                                                         constructingBodyWithBlock:^(id<AFMultipartFormData> formData) {
                                                                                  
        [formData appendPartWithFileData:fileData
                                    name:@"file"
                                fileName:@"file.jpg"
                                mimeType:@"image/jpeg"];
                                                                                  
    } error:nil];
    
    NSURLSessionUploadTask *task;
    task = [self.backgroundNetwork uploadTaskWithStreamedRequest:request
                                                        progress:nil
                                               completionHandler:^(NSURLResponse * __unused response, id responseObject, NSError *error) {
        
    }];

    [task resume];
}
```

**Comment by cesarvilera** [comment]
Turns out the problem with the Content-Length was the proxy. Was adding a blank character at the end of each line.

**Comment by blixt** [comment]
@mattt I've been unable to reproduce this error since I changed the code to keep a reference to both the managers and the `NSProgress` instance while an upload/download is in progress.

**Comment by mattt** [comment]
@blixt Glad to hear it. Thanks for following up.

**Comment by tsheaff** [comment]
I'm still experiencing this issue right now -- it's hard to repro, but it's happening to about ~1% of my users right now.

Here's the stack trace:
```
Foundation	
NSKeyValuePushPendingNotificationPerThread + 340
2
Foundation	
NSKeyValueWillChange + 560
3
Foundation	
-[NSObject(NSKeyValueObserverNotification) willChangeValueForKey:] + 224
4
Foundation	
-[NSProgress _setValueForKeys:settingBlock:] + 292
5
Foundation	
-[NSProgress setTotalUnitCount:] + 244
6
Foundation	
+[NSProgress progressWithTotalUnitCount:] + 116
7
calmapp	
AFURLSessionManager.m line 127
-[AFURLSessionManagerTaskDelegate init]
8
calmapp	
AFURLSessionManager.m line 565
-[AFURLSessionManager addDelegateForDownloadTask:progress:destination:completionHandler:]
9
calmapp	
AFURLSessionManager.m line 737
-[AFURLSessionManager downloadTaskWithRequest:progress:destination:completionHandler:]
10
calmapp	
Download.m line 54
-[Download start]
11
calmapp	
DownloadManager.m line 55
-[DownloadManager downloadFromRemote:withDestinationURL:andProgressUpdate:andCompletion:]
12
calmapp	
ManifestEntry.m line 56
-[ManifestEntry install:]
13
calmapp	
ProgramsManager.m line 289
-[ProgramsManager installStep:forced:withCompletion:]
14
calmapp	
ProgramsManager.m line 270
-[ProgramsManager installStep:withCompletion:]
15
calmapp	
StepListViewController.m line 336
-[StepListViewController selectAndInstallStepAtIndexPath:]
```

and the relevant code snippet:

```
- (void)start
{
    if (self.downloadTask) {
        return;
    }
    
    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    self.sessionManager = [[AFURLSessionManager alloc] initWithSessionConfiguration:configuration];
    
    NSProgress *progress;
    NSURLRequest *request = [NSURLRequest requestWithURL:self.remoteURL];
    __weak typeof(self) weakSelf = self;
    self.downloadTask = [self.sessionManager downloadTaskWithRequest:request
                                                            progress:&progress
                                                         destination:^NSURL *(NSURL *targetPath, NSURLResponse *response)
                         {
                             return [weakSelf generateDestinationURL];
                         }
                                                   completionHandler:^(NSURLResponse *response, NSURL *filePath, NSError *error)
                         {
                            // completion block here
                         }];
    
    self.progress = progress;
    [self addProgressObserver];
    
    [self.downloadTask resume];
}
```

`sessionManager`, `downloadTask` and `progress` are all strongly retained properties.

Thoughts @mattt @blixt , can you guys spot an issue with my implementation?

**Comment by tsheaff** [comment]
As a point of aside, this is only occuring on iOS 8 @blixt did you ever investigate this deeper or see it occur again?

**Comment by tsheaff** [comment]
Turns out there was one (rare) case when the `Download` (`self` in example above) was never calling `removeObserver:forKeyPath:context:`.  I made sure to always call `removeObserver:forKeyPath:context:` exactly once, and now the issue is fixed.

This issue is properly closed, but for anyone coming along this in the future, make sure you properly handle killing KVO on the NSProgress in dealloc of the observer class, because calling it in the completion block of the download task is not necessarily sufficient.
